doc/dev/kdb-record.md

```diff
@@ -0,0 +1,110 @@
+
+The key `user:/record-elektra/config/current_session` contains the name of the current session. If this key does not exist (or its value is empty) the recording plugin is disabled.
+
+All recording sessions are located under `user:/record-elektra/sessions/<SESSION_NAME>`.
+
+Every modification will include its UTC timestamp and the modified key in its path. 
+For example, `user:/record-elektra/sessions/MySession/12345678/system:\/hello`. The value of this key will be the modification action, i.e. `created`, `modified`, `deleted`.
+
+Based on which modification took place (creation, modification or deletion), the old and new value for the key will also be recordes as `.../old` and `.../new`.
+
+For the workflow example from below, it would look like this (11111, 22222, 33333 simulate time stamps):
+```
+user:/record-elektra/sessions/AddNewHosts/111111/system:\/hosts\/ipv4\/www.google.com (= modified)
+user:/record-elektra/sessions/AddNewHosts/111111/system:\/hosts\/ipv4\/www.google.com/old (= 1.2.3.4)
+user:/record-elektra/sessions/AddNewHosts/111111/system:\/hosts\/ipv4\/www.google.com/new (= 8.8.8.8)
+user:/record-elektra/sessions/AddNewHosts/222222/system:\/hosts\/ipv4\/www.microsoft.com (= created)
+user:/record-elektra/sessions/AddNewHosts/222222/system:\/hosts\/ipv4\/www.microsoft.com/new (= 4.4.4.4)
+user:/record-elektra/sessions/AddNewHosts/333333/system:\/hosts\/ipv4\/www.apple.com (= deleted)
+user:/record-elektra/sessions/AddNewHosts/333333/system:\/hosts\/ipv4\/www.apple.com/old (= 9.9.9.9)
+``` 
+
+### Recording plugin
+
+When active, the recording plugin needs to be called for every `kdbGet()` and `kdbSet()` operation. 
+The plugin needs to track which keys have been read and which have been set. 
+
+The plugin MUST ignore `kdbGet()` and `kdbSet()` invocations for its own keys (`user:/record-electra/`).
+
+The `kdbSet()` function of the plugin has to be called in the `POSTCOMMIT` phase of the key database.
+
+As for the execution of the recording plugin, there are multiple options:
+* hardcode the invocation in the `kdbGet()` and `kdbSet()` functions in `kdb.c` like the `gopts` plugin
+* automatically mount the plugin in every backend
+* generify the current plugin system so that global plugins can be defined like backend-specific plugins   
+
+A decision has not yet been made. As the plugin is pretty self-contained, the choice should not effect the implementation of the plugin per se.
+
+[QUESTION:] Is a new instance for a plugin loaded for every backend, or do the backends share a single instance of a plugin?
+
+### Tooling
+
+The tooling will be implemented as part of the `kdb` command line utility.
+
+The most complex part of it will be the `kdb record export` command. 
+When invoked, it must compute the minimal changeset of the session.
```

Comment:
You don't record the full initial state of a session, which is fine, since making a copy of the full KDB for every recording would be pretty expensive. But this means that you only know about keys that were modified or deleted, unmodified keys will simply be missing from the changeset (which is totally normal).

But this means you need a special format for the changeset. The standard storage formats (like `dump`) have no way to record deleted keys. A deleted key would simply not be present. But as stated above, unmodified keys would also be missing. So there is no way to distinguish them.

With the standard `kdb import` you then only have two options:
1. Keep the KDB as is and only overwrite what is contained in the imported KeySet
2. Delete the part of the KDB where data will be imported and write the imported KeySet there

Therefore, with option 1 you might keep keys that should be deleted and with option 2 you might delete keys that should be kept unmodified.

So, you need a somewhat special export format, that can either be imported with a certain plugin, or with a separate `kdb record import` plugin. Your export format needs to record deleted keys and that is not something the normal storage formats are designed to do.

However, it might be enough to use a format that supports metadata (like `dump` or `quickdump`) and add a special metakey (e.g. `meta:/elektra/deleted`) to keys that should be deleted. You'll still need a separate `kdb record import` or a special `kdb import --strategy` to correctly apply these changes though.

Reply:
Good point.

I'm thinking we should define two new functions that plugins can implement - `diffImport` and `diffExport` (names are up for discussion). 

This would allow us to implement various formats in the future.

`kdb record export` and `kdb record import` would then just call the respective function of the plugin.